File: bindings/imgui_bundle/demos/haikus/haiku_implot_heart.py
import time
import numpy as np
from imgui_bundle import implot, imgui_knobs, imgui, run


vals = np.arange(0, np.pi * 2, 0.01)
x = np.power(np.sin(vals), 3) * 16
y = 13 * np.cos(vals) - 5 * np.cos(2 * vals) - 2 * np.cos(3 * vals) - np.cos(4 * vals)

phase = 0
t0 = time.time() + 0.2
heart_pulse_rate = 80


def gui():
    global heart_pulse_rate, phase, t0
    t = time.time()
    phase += (t - t0) * heart_pulse_rate / (np.pi * 2)
    k = 0.8 + 0.1 * np.cos(phase)
    t0 = t
    imgui.text("Bloat free code")
    implot.begin_plot("Heart")
    implot.plot_line("", x * k, y * k)
    implot.end_plot()
    _, heart_pulse_rate = imgui_knobs.knob("Pulse", heart_pulse_rate, 30, 180)


if __name__ == "__main__":
    run(gui, window_size=(300, 450), window_title="Hello!", with_implot=True, fps_idle=0)

File: bindings/imgui_bundle/demos_cpp/demos_immapp/haiku_implot_heart.cpp
#ifdef IMGUI_BUNDLE_WITH_IMPLOT
#include "imgui.h"
#include "implot/implot.h"
#include "imgui-knobs/imgui-knobs.h"
#include "immapp/immapp.h"

#include <cmath>

std::vector<double> VectorTimesK(const std::vector<double>& values, double k)
{
    std::vector<double> r(values.size(), 0.);
    for (size_t i = 0; i < values.size(); ++i)
        r[i] = k * values[i];
    return r;
}

int main(int , char *[]) {
    // Fill x and y whose plot is a heart
    double pi = 3.1415926535;
    std::vector<double>  x, y; {
        for (double t = 0.; t < pi * 2.; t += 0.01) {
            x.push_back(pow(sin(t), 3.) * 16.);
            y.push_back(13. * cos(t) - 5 * cos(2. * t) - 2 * cos(3. * t) - cos(4. * t));
        }
    }
    // Heart pulse rate and time tracking
    double phase = 0., t0 = ImmApp::ClockSeconds() + 0.2;
    float heart_pulse_rate = 80.;

    auto gui = [&]() {
        // Make sure that the animation is smooth
        HelloImGui::GetRunnerParams()->fpsIdling.enableIdling = false;

        double t = ImmApp::ClockSeconds();
        phase += (t - t0) * (double)heart_pulse_rate / (pi * 2.);
        double k = 0.8 + 0.1 * cos(phase);
        t0 = t;

        ImGui::Text("Bloat free code");
        auto xk = VectorTimesK(x, k), yk = VectorTimesK(y, k);
        ImPlot::BeginPlot("Heart", ImmApp::EmToVec2(21, 21));
        ImPlot::PlotLine("", xk.data(), yk.data(), (int)xk.size());
        ImPlot::EndPlot();

        ImGuiKnobs::Knob("Pulse", &heart_pulse_rate, 30., 180.);
    };

    ImmApp::Run(
        gui, "Hello!",
        /*windowSizeAuto=*/false , /*windowRestorePreviousGeometry==*/false, /*windowSize=*/{300, 450},
        /*fpsIdle=*/ 25.f, /*withImplot=*/true);
    return 0;
}

#else // #ifdef IMGUI_BUNDLE_WITH_IMPLOT
#include <cstdio>
int main(int , char *[]) { printf("This demo requires ImPlot\n"); return 0; }
#endif